spec/requests.cddl

```diff
@@ -66,13 +66,21 @@ query-content = {
 query-response = tagged<{
   status: "replied"
   reply: call-reply
+  node_signatures: [* node-signature-of-query-response]
```

Comment:
```suggestion
  ? node_signatures: [* node-signature-of-query-response]
```
Maybe? Or will now all query calls be signed? Then
```suggestion
  node_signatures: [+ node-signature-of-query-response]
```
if that's valid CBOR.

Reply:
Do we really need an optional vector here if the field is omitted? Doesn't CBOR default to an empty vector in that case?

Reply:
I think CBOR distinguishes between field absent, and field present with empty vector as the list. 

The question is what we want: are signatures optional? Or will they be mandatory once this lands?

Reply:
I think if we don't include the certificate in the response (which sounds like a reasonable thing to do as per the discussion [here](https://github.com/dfinity/interface-spec/pull/163/files#r1228516513)) we might be able to afford making the signatures mandatory. This sounds like a nice thing from a security perspective, while it would also enable us to keep the API simple without the need to introduce a second endpoint for signed queries. WDYT?

Reply:
> we might be able to afford making the signatures mandatory [...] without the need to introduce a second endpoint for signed queries

Even with a single endpoint, we can still keep signatures optional and include them based on the request type (this was @rumenov 's suggestion above).

Hence, I see three options (no certificate in the response in all cases):
1. same endpoint, mandatory signatures
2. same endpoint, signatures based on a new field in the request
3. two endpoints, mandatory signatures for the new endpoint

Could we go for option 1? @rumenov

Reply:
> Even with a single endpoint, we can still keep signatures optional and include them based on the request type

Agreed. What I meant to say is that if we want an API thats as simple as possible, I'd rather go for a mandatory signature than for a second endpoint. If we don't mind changing the request type, we can of course have a single endpoint with an optional signature.

Reply:
@mraszyk my preference is option 2 actually. I think David is also advocating for this. Not sure what other think. @nomeata ?

Reply:
I apologize, I wrongly mistyped option 2 initially, sorry about that. 

I am all for **making them mandatory** especially if we think that long term most queries should require signatures.

Reply:
Based on the above, I've specified the list of node signatures to be non-empty.

Reply:
I think there are three aspects to think about whether to make node signatures mandatory (taking both replica-signed queries and certified queries into account).

1. Some queries are already secured by certified variables, e.g. http asset, rosetta.
2. Some queries will count on the certified query feature .
3. Our long-term perspective on the question if all queries should by default require node signatures.

For 1, it will be a waste of work(signing, network traffic, merging certificate, merging sigs) for these queries.
For 2, NNS dapp has asked for this feature and it will be the main user in the early stage. I don't foresee a surge of demand for certified queries. 
For 3, I am not sure that all requests should by default demand signatures if we considering charging queries in the future.

For the reasons above, I go for option 2. WDYT? @derlerd-dfinity  @mraszyk @rumenov @nomeata

Reply:
Your comments on the other thread invalidated my claim that it is likely the most reasonable thing to do to not include the certificate into the response. So I'm also leaning towards Option 2 (also see [here](https://github.com/dfinity/interface-spec/pull/163#discussion_r1232046914)).

Reply:
There are two separate decisions to be taken:

A. should we include node signatures in the response?
(1) same endpoint, mandatory signatures
(2) same endpoint, signatures based on a new field in the request
(3) two endpoints, mandatory signatures for the new endpoint

B. should we include (node key) certificate in the response?
(1) never (=> must make a read_state request)
(2) based on an optional integer field included in the request
(3) always (=> no need for a separate read_state request)

I believe we had already converged to A1 and stay with that decision.

Regarding B, we decided B1 and now we're changing our decision to B2, correct?

Reply:
> I believe we had already converged to A1 and stay with that decision.

B2 with A1 is an odd choice. If you already introduce configurabilityim the request, then at least I don't see why not allowing the option “no signatures wanted”, and thus A2 (which would be my favorite combination). Maybe this has all been discussed to death internally already, though :-)

Reply:
> I believe we had already converged to A1 and stay with that decision.
> [...]
> > B2 with A1 is an odd choice.

I agree. that it is an odd choice. I think we should either make both signatures and certificates requestable or nothing.

In my understanding we had converged on making signatures mandatory under the [assumption](https://github.com/dfinity/interface-spec/pull/163#discussion_r1229169260) that it doesn't make sense to include the certificate, but Shuo's argument [here](https://github.com/dfinity/interface-spec/pull/163#discussion_r1232046914) invalidated this IMHO. 

So I think we should either:
1. Make both, the certificate and the signatures requestable with (an) additional parameter(s) in the request
2. Stick with decision A1 but then also don't include certificates.

With this new input, my tendency is towards 1 and it seems others landed on the same conclusion. If I understood correctly, @rumenov's objections were related to making the API more complicated by allowing the client to select what it wants. @rumenov would you also be fine with 1 given the new inputs linked above?